### PR TITLE
Initialize hash based collections appropriately

### PR DESCRIPTION
**Pull request by mmedenjak**
The hash collections were initialised with a capacity that caused
rehashing even though the final size was known in advance. Fixed
sizing and introduced some minor refactoring.

Rebased version of : https://github.com/hazelcast/hazelcast/pull/9597
Fixes : https://github.com/hazelcast/hazelcast/issues/9445
Thanks to @bokken for submitting the PR

### CODE REVIEW

**Comment by mmedenjak** [comment]
@jerrinot @tombujok do you think this could be merged for 3.9? Or is this too many changes?

**Comment by pveentjer** [comment]
@tombujok @mmedenjak what to do with this PR?

**Comment by mmedenjak** [comment]
@pveentjer I think it's useful. It was proposed from the community but it was abandoned. I took on the job of rebasing it. 
It would be nice to have it but maybe as time is short maybe we postpone it for 3.10

File: hazelcast/src/main/java/com/hazelcast/client/impl/operations/GetConnectedClientsOperation.java [hunk 1]
```diff
@@ -37,8 +38,9 @@ public GetConnectedClientsOperation() {
     @Override
     public void run() throws Exception {
         ClientEngineImpl service = getService();
-        this.clients = new HashMap<String, ClientType>();
-        for (Client clientEndpoint : service.getClients()) {
+        final Collection<Client> serviceClients = service.getClients();
+        this.clients = MapUtil.createHashMap(serviceClients.size());
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/client/impl/operations/GetConnectedClientsOperation.java]: Maybe an idea to create a method 'createHashMapWithSize that accepts a collection as argument so you don't need to call 'size'? Anyhow; no need to change.

**Comment by mmedenjak** [comment]
@tkountis @pveentjer rebased it, used static imports and reverted some files and lines that were changed only in formatting/line break type

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java [hunk 2]
```diff
@@ -29,8 +29,10 @@
 
 final class WaitersInfo implements IdentifiedDataSerializable {
 
+    private static final int INITAIL_WAITER_SIZE = 4;
+
     private String conditionId;
-    private Set<ConditionWaiter> waiters = new HashSet<ConditionWaiter>(2);
+    private Set<ConditionWaiter> waiters = new HashSet<ConditionWaiter>(INITAIL_WAITER_SIZE);
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java]: The value has changed from 2 to 4, is this intentional?

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheProxyBase.java [hunk 3]
```diff
@@ -61,6 +62,12 @@
 
     private static final int TIMEOUT = 10;
 
+    /**
+     * Speculative factor to be used when initialising collections
+     * of an approximate final size.
+     */
+    private static final double SIZING_FUDGE_FACTOR = 1.3;
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractCacheProxyBase.java]: I think we have this factor and formula three times in the code now. Should we put this into one of the utility classes?

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java (hunk 2 above)
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java]: I will check that. Actually I wanted to go through the whole PR once more as I initially based it on a community PR. The guy put in an effort to understand parts of the code and try and guess the size on each change.

**Comment by pveentjer** [comment]
@mmedenjak how to move forward? There are 2 thumbs up.

**Comment by mmedenjak** [comment]
@pveentjer I'll probably merge it, I just want to go over each change again (and answer David's question)

**Comment by mmedenjak** [comment]
@pveentjer moving this to 3.9.1 as we are close to releasing 3.9 and I don't want to "muddy the waters" with this PR now

**Comment by jerrinot** [comment]
@mmedenjak can you please rebase?

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java (hunk 2 above)
Comment (Donnerbart) [reply hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java]: BTW, there is a typo in the constant name :) `INITAIL_WAITER_SIZE` should be `INITIAL_WAITER_SIZE` :)

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/concurrent/lock/WaitersInfo.java]: Changing back to 2 and using the `com.hazelcast.util.SetUtil#createHashSet` method.

File: hazelcast/src/main/java/com/hazelcast/concurrent/semaphore/SemaphoreService.java [hunk 4]
```diff
@@ -139,7 +139,7 @@ public void beforeMigration(PartitionMigrationEvent partitionMigrationEvent) {
 
     @Override
     public Operation prepareReplicationOperation(PartitionReplicationEvent event) {
-        Map<String, SemaphoreContainer> migrationData = new HashMap<String, SemaphoreContainer>();
+        Map<String, SemaphoreContainer> migrationData = createHashMap(containers.size() / 2);
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/concurrent/semaphore/SemaphoreService.java]: Reverting this as it assumes half of the containers will be replicated which does not need to be true.

File: hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java [hunk 5]
```diff
@@ -773,7 +773,7 @@ protected void getAllInternal(Set<K> keys, List<Data> dataKeys, List<Object> res
             return Collections.emptyMap();
         }
 
-        Map<Integer, List<Data>> idToKeys = new HashMap<Integer, List<Data>>();
+        Map<Integer, List<Data>> idToKeys = createHashMap(partitionService.getPartitionCount());
```
Comment (mmedenjak) [inline hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java]: I'm going to revert this as we don't know that there will be enough keys to be distributed on all partitions. Also, we are planning on changing the default partition count to 1999

**Comment by mmedenjak** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/cache/impl/CacheEventListenerAdaptor.java [hunk 6]
```diff
@@ -185,7 +187,7 @@ private void handleEvent(int type, Collection<CacheEventData> keys) {
     }
 
     private Iterable<CacheEntryEvent<? extends K, ? extends V>> createCacheEntryEvent(Collection<CacheEventData> keys) {
-        HashSet<CacheEntryEvent<? extends K, ? extends V>> evt = new HashSet<CacheEntryEvent<? extends K, ? extends V>>();
+        Set<CacheEntryEvent<? extends K, ? extends V>> evt = createHashSet(keys.size());
```
[the remark this answers is not in the document]
Comment (bokken) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/CacheEventListenerAdaptor.java]: but that is the worst case scenario, right? Resizing a hash set is a relatively expensive operation in both memory and cpu (especially as they grow in size). Oversizing is only expensive in memory (and indeed may have some benefits in cpu as it reduces probability of hash collisions).

Perhaps a compromise which at least guarantees worst case is 1 resize:
createHashSet(Math.max(keys.size() / 2, 8));

File: hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetAllOperation.java [hunk 7]
```diff
@@ -49,14 +51,15 @@ public CacheGetAllOperation(String name, Set<Data> keys, ExpiryPolicy expiryPoli
...
+        keys = new HashSet<Data>();
     }
 
     public void run() {
         ICacheService service = getService();
         ICacheRecordStore cache = service.getOrCreateRecordStore(name, getPartitionId());
 
         int partitionId = getPartitionId();
-        Set<Data> partitionKeySet = new HashSet<Data>();
+        Set<Data> partitionKeySet = createHashSet(keys.size());
```
[the remark this answers is not in the document]
Comment (bokken) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/operation/CacheGetAllOperation.java]: same comment as above about dealing with worst case scenario. and at least reducing the cost/impact of that worst case.

**Comment by mmedenjak** [comment]
@bokken I am merging this PR as-is since it's pretty big and it has been dragged on for pretty long enough. I will then make a new PR with the other suggestions + one refactoring

**Pull request merged**